Add day to date on activity page sidebar

hmstepanek opened the pull request:
Add day to group created date on sidebar so date format is now:
Month day, year aka: August 1, 2018. Some tests were updated to
expect this new date format. Note the date format was changed from
using the c library strftime to using the python format method as
the c library does not support %-d on all systems (a way of removing
the default zero padding of the day). This is a fix for 
https://github.com/hypothesis/client/issues/644.

hmstepanek commented:
In response to @robertknight earlier comment I didn't really want to get into the whole locale thing right now-was kinda hoping we could table that for the future? I didn't mean to open a can of worms here. My thinking was the date is already not guaranteed to be in the user's time zone and it's not like the user isn't able to distinguish between what the month/day/year is since the month is spelled out so no harm? I agree with what you are saying about the locale thing but I was hoping that could be a separate issue for the future?

dawariley commented:
From a user's point of view I would agree that having the month/day/year US date format isn't a deal breaker for now and we should proceed with that and build out the locale function in the future.

When I initially added this requirement to https://github.com/hypothesis/client/issues/644, it came from a request from Dan and I raised the location specific date format and we both agreed that it wouldn't be a big deal for v1.

seanh commented:
Before:

![before-group](https://user-images.githubusercontent.com/22498/36734817-8c486cce-1bcc-11e8-91bb-d04e2c7b1a64.png)

![before-user](https://user-images.githubusercontent.com/22498/36734818-8c67a300-1bcc-11e8-8fa8-88fe5c3688e6.png)

After:

![after-group](https://user-images.githubusercontent.com/22498/36734829-93107f1a-1bcc-11e8-93df-5c1c824187d7.png)

![after-user](https://user-images.githubusercontent.com/22498/36734831-93bcfbb4-1bcc-11e8-9f39-092ec9fb2e0c.png)

seanh commented on h/views/activity.py:
```diff
@@ -24,6 +24,10 @@
 
 
+def _format_date(date):
+    return "{d:%B} {d.day}, {d:%Y}".format(d=date)
```
I've never seen this way of formatting dates in Python! Very interesting. So `{d:%B}` produces numbers like `28` and `1` (as opposed to `01`)? And there's no way to do that with `strftime()`?

Anyway, I think you should mover `_format_date()` into a utility module, maybe [h.util.datetime](https://github.com/hypothesis/h/blob/master/h/util/datetime.py), and add some unit tests to it. There's probably not _that_ much to test - but at least test that it works with both double- and single-digit days, and with both date and datetime objects. If nothing else this will prevent someone from coming along and replacing with with a more typical usage of `strftime()` later, and not realising what they're breaking.

robertknight replied:
Heh, I wasn't aware of this either. Any pros/cons compared to `strftime` we should be aware of?

hmstepanek replied:
I added a comment about this in the commit message. We can't use strftime to produce a non zero padded day as we would use %-d and that isn't supported on all systems. Note strftime calls into the c library of the os and %-d is a 1999 standard where as apparently our qa system (not sure about prod) is 1998 standard and doesn't support it.

robertknight replied:
Thanks, that's interesting. Further discussion: https://hypothes-is.slack.com/archives/C4K6M7P5E/p1519755955000065

I verified that `%-d%` indeed doesn't work in an h Docker container, so this is more portable.

robertknight commented:
> My thinking was the date is already not guaranteed to be in the user's time zone and it's not like the user isn't able to distinguish between what the month/day/year is since the month is spelled out so no harm?

Providing that there is a reasonably low probability of the changes confusing non-US/UK users compared to the existing formatting then I would say that we could revisit i18n issues later.

I'll give an example of where blinding using US formatting would be a mistake though: Supposing the date was currently written as YYYY-MM-DD and you changed it to MM/DD/YYYY which is common in the US. This could be confusing in other parts of the world (eg. the UK) where the date is usually written DD/MM/YYYY. As you say, we are spelling out the month name in full here, so that doesn't apply.

In the context of number formatting, a comma is used as a _thousands_ separator in the US/UK but it is used as a _decimal_ separator elsewhere, so that gave me pause for thought.

hmstepanek commented:
I would like to move that function into the util module per @seanh 's suggestion before merging this, if that's ok. I'll do that tomorrow AM-ran out of time today.

dawariley commented:
"In the context of number formatting, a comma is used as a thousands separator in the US/UK but it is used as a decimal separator elsewhere, so that gave me pause for thought."

Is it possible to be 'comma agnostic' and just leave it out altogether?

robertknight commented:
> Is it possible to be 'comma agnostic' and just leave it out altogether?

It was omitted previously, but grouping does make larger numbers easier to read so I'm OK with adding it.

There is a library we can use to format numbers with grouping according to the conventions of the user's language & region. So we could use a comma for now and localize it in follow-up work.

What is needed from a design perspective is a decision on whether to display numbers with group separators or not.

dawariley commented:
ok! sounds good. by group separators do you me '-' vs '/'.

I think the preference is to retain the 'natural language' version of: "August 1, 2018."

If I have your ask wrong.. you might need to define 'group separator' for me.

robertknight commented:
> ok! sounds good. by group separators do you me '-' vs '/' ?

My comment about group separators was referring to general _number_ rather than _date_ formatting as in Hannah's other PR, it doesn't apply here.

dawariley commented:
Let's follow the 'Google' standard: "periods as a thousands separator and commas for decimals", factoring in locale when we have time.

segdeha commented:
So, `10.000,00`? That would look broken to our U.S. users who make up most of our user base.

hmstepanek commented:
Well let’s be fair, there are no partial annotations so it would look like 10.000. We could also use spaces potentially.

robertknight commented:
> Let's follow the 'Google' standard: "periods as a thousands separator and commas for decimals", factoring in locale when we have time.

This discussion pertains to a different PR, let's not hash it out here. This PR is just about the date formatting change.

hmstepanek commented:
Ok I have moved the date format method into the util location that @seanh requested. Also changed the name to utc_us_style_date. If @seanh feels good about this change I'll go ahead and merge it?

robertknight commented on tests/h/util/datetime_test.py:
```diff
@@ -28,3 +28,7 @@ def test_utc_iso8601():
     t = datetime.datetime(2016, 2, 24, 18, 3, 25, 7685, Berlin())
     assert utc_iso8601(t) == '2016-02-24T18:03:25.007685+00:00'
+
+def test_us_style_date():
```
Nitpick - This should be `test_utc_us_style_date`.

robertknight commented:
Looks like there is a lint failure and I had a tiny suggestion regarding a test name change. Once those are resolved, I think you can go ahead and merge this.

hmstepanek commented:
oopsy...i fix.
